pkg/api/dtos/models_test.go

```diff
@@ -3,11 +3,71 @@ package dtos
 import (
 	"testing"
 
+	"github.com/grafana/grafana/pkg/components/simplejson"
 	"github.com/grafana/grafana/pkg/services/user"
 	"github.com/grafana/grafana/pkg/setting"
 	"github.com/stretchr/testify/assert"
 )
 
+func TestGetUniqueDatasourceTypes(t *testing.T) {
+	testcases := []struct {
+		desc    string
+		queries []*simplejson.Json
+		result  []string
+	}{
+		{
+			desc:   "can get unique datasource names",
+			result: []string{"prometheus", "mysql"},
+			queries: []*simplejson.Json{
+				simplejson.NewFromAny(map[string]interface{}{
+					"datasource": map[string]interface{}{
+						"type": "prometheus",
+						"uid":  "uid1",
+					},
+				}),
+				simplejson.NewFromAny(map[string]interface{}{
+					"datasource": map[string]interface{}{
+						"type": "prometheus",
+						"uid":  "uid2",
+					},
+				}),
+				simplejson.NewFromAny(map[string]interface{}{
+					"datasource": map[string]interface{}{
+						"type": "mysql",
+						"uid":  "uid3",
+					},
+				}),
+			},
+		},
+		{
+			desc:   "returns empty slice when datasources have no type property",
+			result: []string{},
+			queries: []*simplejson.Json{
+				simplejson.NewFromAny(map[string]interface{}{
+					"datasource": map[string]interface{}{
+						"uid": "uid1",
+					},
+				}),
+				simplejson.NewFromAny(map[string]interface{}{
+					"datasource": map[string]interface{}{
+						"uid": "uid3",
+					},
+				}),
+			},
+		},
```

Comment:
Should we add a test for an empty datasource? Not sure if it could be a valid case.

Reply:
I don't think it's a valid case. Not sure how it would happen unless someone manually edits the dashboard json, which would break the query anyways. I feel once we have a MetricRequest, we should expect it has the content it says it should. Whatever creates it should be responsible for making sure the queries have the right structure. Thoughts?

Reply:
Yes, if it is not a valid case it should not be covered.